### DISCUSSION TITLE
Why does `extends` in `tsconfig.json` not work with local packages?

### DISCUSSION THREAD

**Discussion by PaulRBerg** [cypress-io/cypress#27054] [Questions and Help] [answered]
As per https://github.com/cypress-io/cypress/issues/5674, it looks like Cypress should support the `extends` functionality in `tsconfig.json`.

I noticed that relative paths work:

```json
{
  "extends": "../../paths.json",
}
```

But the following doesn't:

```json
{
    "extends": "@org-name/tsconfig/paths.json"
}
```

Where `@org-name` is the name of our `package.json` organization name, in a local yarn-based monorepo.

Why doesn't this work? We're getting the following error:

```text
Error: Webpack Compilation Error
./cypress/e2e/deployed.cy.ts
Module not found: Error: Can't resolve '~/labs/constants' in '/Users/prb/Sablier/v2-interfaces/apps/labs/cypress/e2e'
resolve '~/labs/constants' in '/Users/prb/Sablier/v2-interfaces/apps/labs/cypress/e2e'
  Parsed request is a module
  using description file: /Users/prb/Sablier/v2-interfaces/apps/labs/package.json (relative path: ./cypress/e2e)
    Field 'browser' doesn't contain a valid alias configuration
    Looked for and couldn't find the file at the following paths:
```

Note that that shared TSConfig file defines paths, e.g.

```json
{
  "extends": "./base.json",
  "compilerOptions": {
    "baseUrl": "../../",
    "paths": {
      "~/labs/*": ["apps/labs/src/*"],
    }
  }
}
```

See the full log [here](https://app.warp.dev/block/YnW2NJRJBuK7ENfEeiU004).

**Comment by portyu9** [comment, marked as answer]
Hi @PaulRBerg 

The error shown here is not evidence that TypeScript's package-based `extends` is unsupported.

The important part of the error is:

```text
Webpack Compilation Error
Module not found: Error: Can't resolve '~/labs/constants'
```

Your shared `tsconfig` defines `~/labs/*` under `compilerOptions.paths`, but **Webpack does not automatically use TypeScript `paths` mappings**. Those mappings tell TypeScript how to resolve/type-check imports; the bundler needs matching resolution configuration as well.

For Cypress E2E specs using `@cypress/webpack-preprocessor`, the current supported approach is `tsconfig-paths-webpack-plugin`:

```js
const { defineConfig } = require('cypress')
const webpackPreprocessor = require('@cypress/webpack-preprocessor')
const TsconfigPathsPlugin = require('tsconfig-paths-webpack-plugin')

module.exports = defineConfig({
  e2e: {
    setupNodeEvents(on, config) {
      const options = webpackPreprocessor.defaultOptions

      options.webpackOptions.resolve = {
        ...options.webpackOptions.resolve,
        plugins: [new TsconfigPathsPlugin()],
      }

      on('file:preprocessor', webpackPreprocessor(options))
    },
  },
})
```

It is also worth separating the two questions by running:

```sh
npx tsc -p path/to/your/cypress/tsconfig.json --showConfig
```

If that command cannot resolve:

```json
"extends": "@org-name/tsconfig/paths.json"
```

then the problem is the Yarn/workspace package resolution or the package itself.

If `tsc --showConfig` succeeds and shows the expected `paths`, but Cypress still reports `Webpack Compilation Error`, then the problem is bundler resolution and `tsconfig-paths-webpack-plugin` is the relevant fix.

Also, the linked #5674 is not about this behavior: that issue requested an `extends` feature for Cypress's own configuration (`cypress.json` at the time), so it does not establish that Webpack should resolve `tsconfig` aliases.

Please mark this answer as accepted if it helped, thank you.
